File: 279-perfect-squares/279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        squares = [num**2 for num in range(1, int(sqrt(n)) + 1)]
        
        q = {n}
        ans = 0
        while q:
            nq = set()
            for remaining in q:
                if remaining == 0:
                    return ans
                
                for j in range(int(sqrt(remaining))):
                    nr = remaining - squares[j]
                    if nr not in nq:
                        nq.add(nr)
            q = nq
            ans += 1
        return ans

File: 279-perfect-squares/279-perfect-squares.java
class Solution {
    public int numSquares(int n) {
        int m = (int) Math.sqrt(n) + 1;
        int[] squares = new int[m];
        for(int num = 1; num < m; num++)
            squares[num - 1] = num * num;
        
        HashSet<Integer> q = new HashSet<>();
        q.add(n);
        
        int ans = 0;
        while(q.size() > 0){
            HashSet<Integer> nq = new HashSet<>();
            for(int remaining : q){
                if(remaining == 0)
                    return ans;
                
                int r = (int) Math.sqrt(remaining);
                for(int i = 0; i < r; i++){
                    int nr = remaining - squares[i];
                    if(!nq.contains(nr))
                        nq.add(nr);
                }
            }
            q = nq;
            ans++;
        }
        
        return ans;
    }
}